src/lib.rs

```diff
@@ -119,40 +110,299 @@ impl Ascii {
 
 impl fmt::Display for Ascii {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Display::fmt(&(self.chr as char), f)
+        self.as_char().fmt(f)
     }
 }
 
-impl fmt::Display for Vec<Ascii> {
+impl fmt::Debug for Ascii {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Display::fmt(&self[..], f)
+        self.as_char().fmt(f)
+     }
+}
+
+#[derive(Clone)]
+pub struct AsciiString {
+    vec: Vec<Ascii>,
+}
+
+impl AsciiString {
+    pub fn new() -> AsciiString {
+        AsciiString { vec: Vec::new() }
+    }
+
+    pub fn from_bytes<B>(bytes: B) -> Result<AsciiString, B> where B: Into<Vec<u8>> + AsRef<[u8]> {
+        if bytes.as_ref().is_ascii() {
+            unsafe { Ok( AsciiString::from_vec(bytes.into()) ) }
+        } else {
+            Err(bytes)
+        }
+    }
+
+    unsafe fn from_vec(src: Vec<u8>) -> AsciiString {
+        let vec = Vec::from_raw_parts(src.as_ptr() as *mut Ascii,
+                                      src.len(),
+                                      src.capacity());
+
+        // We forget `src` to avoid freeing it at the end of the scope.
+        // Otherwise, the returned `AsciiString` would point to freed memory.
+        mem::forget(src);
+        AsciiString { vec: vec }
     }
 }
 
-impl fmt::Display for [Ascii] {
+impl Deref for AsciiString {
+    type Target = AsciiStr;
+
+    fn deref<'a>(&'a self) -> &'a AsciiStr {
+        unsafe { mem::transmute(&self.vec[..]) }
+    }
+}
+
+impl DerefMut for AsciiString {
+    fn deref_mut<'a>(&'a mut self) -> &'a mut AsciiStr {
+        unsafe { mem::transmute(&mut self.vec[..]) }
+    }
+}
+
+impl PartialEq for AsciiString {
```

Comment:
Yes, also `Eq`, `PartialOrd`, `Ord` and `Hash` can be derived on `AsciiString`. On `AsciiStr` only `Hash` can be derived because of https://github.com/rust-lang/rust/issues/23649.